### PR TITLE
Plugins management with Flux architecture

### PR DESCRIPTION
**Pull request by koke**
Introduces the new Flux-inspired data flow architecture. There's a good overview in the WordPressFlux project [README](https://github.com/wordpress-mobile/WordPress-iOS/blob/72884cfb3a84b0eb365f9e7bd356a3249b47c44c/WordPressFlux/README.md), and I've temporarily pushed it to my site along with the code documentation: [WordPressFlux - A Flux-inspired data flow architecture](https://koke.me/WordPressFlux/index.html)

I wanted to make an example project, but I think this initial PR is already large enough, we can do that later.

Another odd thing that I've added to this PR is [CredentialsService](https://github.com/wordpress-mobile/WordPress-iOS/blob/72884cfb3a84b0eb365f9e7bd356a3249b47c44c/WordPress/Classes/Services/CredentialsService.swift) and [JetpackSiteRef](https://github.com/wordpress-mobile/WordPress-iOS/blob/72884cfb3a84b0eb365f9e7bd356a3249b47c44c/WordPress/Classes/Models/JetpackSiteRef.swift).
Passing a `siteID: Int` around to refer to a site is not enough: while uncommon, the same site could exist twice in the app in multiple accounts, plus the store needs to create a remote and obtain credentials for that site. Passing managed objects around is also not ideal, specially since we don't have them for plugins yet.

So we have `JetpackSiteRef`, which is basically a `siteID`/`username` pair, and `CredentialsService`, which will bypass Core Data, and look directly in the Keychain for credentials.

This isn't necessarily a definitive solution, but it felt like a good lightweight compromise until we decide to move forward with this and start implementing the account and site stores. Once those are in place, there might be a more elegant solution to passing site references and credentials.

### CODE REVIEW

**Comment by koke** [comment]
@oguzkocer thanks for the feedback!

One of my initial goals included what you're describing: having queries subscribe for change events when the relevant data actually changes. There are two parts to this.

## Detecting when the state has actually changed

This part is not hard but it is cumbersome. We need `State` to conform to `Equatable` so we can figure out if the old value is different than the new one when it changes. And you have to implement this comparison for every state type, which means a lot of boilerplate code. Swift 4.1 will make this so much easier with [synthesized Equatable conformance](https://github.com/apple/swift-evolution/blob/master/proposals/0185-synthesize-equatable-hashable.md), but until then it felt like forcing implementers to write too much code for a performance optimization we're not sure we need.

So I postponed this and left it for the implementer to filter out relevant changes. I didn't do it in plugins, but you can definitely filter out non-changes in the view model: `PluginListViewModel.State` is only concerned about loading state and the list of plugins for one site, so it's both easier to write an Equatable implementation for it, and actually run it, since it doesn't care if plugins for other sites have changed.

## Getting only changes relevant to the subscribed query

I spent a lot of time trying to make this work, and like the previous issue, I ended up dropping it as it was more of a performance optimization, and it might be premature. I agree it's a nice feature to have, but after fighting with generics and the compiler, I was reaching the limits of Swift type system.

In this latest implementation, `Query` is an empty protocol and a query can be anything. I opted for "dumb" queries that don't have any logic. My initial idea is that a `Query` would know two things: how/when to make a request given the current store state, and how to calculate the wanted results from the store state:

```swift
public protocol Query {
    associatedtype State
    associatedtype Result: Equatable
    var selector: (State) -> Result { get }
    var request: (State) -> Void { get }
}
```

But when you start adding associated types and generics, you have to specify them everywhere and it adds a lot of coupling between components. At the end, the main issue is that since the `Store` is keeping track of all the queries, you'd have to know what `Result` is to store them since you can't have heterogeneous arrays. So this had the limitation that all the queries for a store would have the same `Result` type, which wasn't good enough (a query for a single plugin and a query for all plugins return different stuff).

What you suggest about adding a payload to `emitChange()` is interesting (maybe you were thinking of `OnSitePluginsChanged` instead?), but it's trickier in practice. I think it makes more sense if we move away from `StatefulStore`, as the point of it was that you just modify the state and let the store deal with that. I have left this out of this iteration, but I'd like to experiment more with reducer stores. But even if you go closer to the approach of FluxC, it is very manual and error prone. For instance, looking at FluxC plugins code, I don't see `OnSitePluginsChanged` being triggered when a plugin is updated individually or deleted. Maybe it's just not implemented yet, but I'd prefer an approach when the framework figures out who needs to get notified about what so the developer doesn't have to.

File: WordPress/Classes/Services/InAppPurchaseStore.swift [hunk 1]
```diff
@@ -47,7 +47,7 @@ enum StoreCoordinatorError: Error {
 /// - `StoreKitCoordinator.TransactionDidFailNotification` on failure.
 ///   The notification's `userInfo` will also contain the productID of the attempted
 ///   purchased product, as well as a localized error message under `NSUnderlyingErrorKey`.
-class StoreCoordinator<S: Store> {
+class StoreCoordinator<S: InAppPurchaseStore> {
```
[the remark this answers is not in the document]
Comment (koke) [reply WordPress/Classes/Services/InAppPurchaseStore.swift]: Yes, sorry, I thought of extracting the previous Store/Action changes to a separate PR but completely forgot about that.

**Comment by koke** [comment]
@aerych I left some ideas about smarter state changes on the [Future Ideas](https://github.com/wordpress-mobile/WordPress-iOS/blob/72884cfb3a84b0eb365f9e7bd356a3249b47c44c/WordPressFlux/README.md#future-ideas) section of the README, but it's up to each store/view model now. If there are actual performance issues, we can investigate solutions, but it seemed a bit premature for a v1.

As for Core Data, I haven't experimented yet. I'd love to see us move towards non-managed models, but I think we could have the stores keep their own managed context and return managed objects, as long as there's a clear expectation that nothing other than the store should modify them. Maybe it's even OK to keep using NSFetchedResultsController for lists 🤷🏽‍♂️

**Comment by SergioEstevao** [comment]
First of all I want to congratulate @koke for an excellent work on this PR, functionality, code and most important documentation and for me the cherry on top is that README.md explaining the architecture

Regarding this specific PR I think the approach of FLUX works pretty well for plugins and for me it's a :shipit: for me.

But as others told before me I think we need to test the approach using CoreData. Looking at the code I think this can be achievable using a Store that encapsulates our current Services. One store could have two services inside one for the mainContext to return results to the VC and another with a background context for interacting with network operations. The services will implement all the actions of the Store, and the store can be responsible for synchronisation and possible error handling.

In terms of updates we could create Query that encapsulates a protocol interface that imitates the NSFetchedResultController event saying what: changes, insert, removes happened on an Entity.

I think this approach will allows us to migrate other parts of the app keeping a lot of the code that we currently have.

I will be interested to try this approach for the async Media work we are doing at the moment.

**Comment by koke** [comment]
> saw this behaviour, but there's no explanation why it's not working

@frosty that's going on the snackbar 😉 😉

**Pull request merged**